private/pkg/storage/storagegit/bucket.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storagegit
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+
+	"github.com/bufbuild/buf/private/pkg/git"
+	"github.com/bufbuild/buf/private/pkg/normalpath"
+	"github.com/bufbuild/buf/private/pkg/storage"
+	"github.com/bufbuild/buf/private/pkg/storage/storageutil"
+)
+
+type bucket struct {
+	objectReader git.ObjectReader
+	symlinks     bool
+	root         git.Tree
+}
+
+func newBucket(
+	objectReader git.ObjectReader,
+	symlinksIfSupported bool,
+	root git.Tree,
+) (storage.ReadBucket, error) {
+	return &bucket{
+		objectReader: objectReader,
+		symlinks:     symlinksIfSupported,
+		root:         root,
+	}, nil
+}
+
+func (b *bucket) Get(ctx context.Context, path string) (storage.ReadObjectCloser, error) {
+	node, err := b.root.Traverse(b.objectReader, normalpath.Components(path)...)
+	if err != nil {
+		if errors.Is(err, git.ErrSubTreeNotFound) {
+			return nil, storage.NewErrNotExist(path)
+		}
+		return nil, err
+	}
+	switch node.Mode() {
+	case git.ModeFile, git.ModeExe:
+		data, err := b.objectReader.Blob(node.Hash())
+		if err != nil {
+			return nil, err
+		}
+		return &namedReader{
+			info:   b.newObjectInfo(path),
+			reader: bytes.NewReader(data),
+		}, nil
+	case git.ModeSymlink:
+		if !b.symlinks {
+			return nil, storage.NewErrNotExist(path)
+		}
+		// Symlinks are stored as blobs that reference the target path as a relative
+		// path. We can follow this symlink trivially.
+		data, err := b.objectReader.Blob(node.Hash())
+		if err != nil {
+			return nil, err
+		}
+		path = normalpath.Join(
+			normalpath.Base(path),
+			normalpath.Normalize(string(data)),
+		)
```

Comment:
Surprised this doesn't return optional error. What if the symlink content is not actually a path but a bunch of nonsense bytes? I guess we'd just return `NotFound` eventually, but I thought we could catch it here.